### ISSUE TITLE
Wallaby not finding node library with jest, but jest works fine

### ISSUE THREAD

**Issue by bbugh** [wallabyjs/public#2050]
### Issue description or question

I'm using the `given` library for testing. I recently rewrote all of our jest specs to TypeScript, which is working great.

Wallaby is saying `TypeError: given is not a function` for all of the TypeScript tests that use it. I tried given in a `spec.js` file and it works fine for Wallaby but breaks Jest. Seems to be something with TypeScript importing libraries, except it works fine with Jest, and TypeScript successfully reports the types that I've declared.

```typescript
import * as given from 'given2'
import { mount } from '@vue/test-utils'
import BaseFormInput from '#/common/components/base/BaseFormInput.vue'

given('wrapper', () => mount(BaseFormInput, {  // Wallaby: TypeError: given is not a function
  propsData: {
    append: given.append,
    field: { $model: given.model },
    prepend: given.prepend,
    type: given.type
  }
}))
```

I've tried a lot of different things to fix it, like manually changing `esModuleInterop: true`, `modules: 'commonjs'`, etc. overriding Wallaby's TypeScript compiler, but that wasn't working either.

I also tried adding `given2` to `transformIgnorePatterns`, that didn't work. 

I'm pretty stumped, any help is appreciated! 

### Wallaby.js configuration file

```javascript
process.env.BABEL_ENV = 'test'

module.exports = wallaby => {
  return {
    files: [
      'app/javascript/**/*',
      '!app/javascript/**/*.spec.js',
      '!app/javascript/**/*.spec.ts',
      'package.json',
      'jest.config.js',
      'tsconfig.json'
    ],

    tests: ['app/javascript/**/*.spec.js', 'app/javascript/**/*.spec.ts'],

    env: {
      type: 'node',
      runner: 'node'
    },

    preprocessors: {
      '**/*.js?(x)': file =>
        require('babel-core').transform(file.content, {
          sourceMap: true,
          compact: false,
          filename: file.path,
          babelrc: true
        })
    },

    setup (wallaby) {
      const jestConfig = require('./package').jest || require('./jest.config')
      delete jestConfig.transform['^.+\\.tsx?$']
      wallaby.testFramework.configure(jestConfig)
    },

    testFramework: 'jest',

    debug: true
  }
}

```

### Other config files

<details>
<summary>Babel config</summary>

```javascript
module.exports = function (api) {
  var validEnv = ['development', 'test', 'production']
  var currentEnv = api.env()
  var isDevelopmentEnv = api.env('development')
  var isProductionEnv = api.env('production')
  var isTestEnv = api.env('test')

  if (!validEnv.includes(currentEnv)) {
    throw new Error(
      'Please specify a valid `NODE_ENV` or ' +
        '`BABEL_ENV` environment variables. Valid values are "development", ' +
        '"test", and "production". Instead, received: ' +
        JSON.stringify(currentEnv) +
        '.'
    )
  }

  return {
    presets: [
      isTestEnv && [
        require('@babel/preset-env').default,
        {
          targets: {
            node: 'current'
          }
        }
      ],
      (isProductionEnv || isDevelopmentEnv) && [
        require('@babel/preset-env').default,
        {
          forceAllTransforms: true,
          useBuiltIns: 'entry',
          modules: false,
          exclude: ['transform-typeof-symbol']
        }
      ],
      '@babel/preset-typescript'
    ].filter(Boolean),
    plugins: [
      require('babel-plugin-macros'),
      require('@babel/plugin-syntax-dynamic-import').default,
      isTestEnv && require('babel-plugin-dynamic-import-node'),
      require('@babel/plugin-transform-destructuring').default,
      [
        require('@babel/plugin-proposal-class-properties').default,
        {
          loose: true
        }
      ],
      [
        require('@babel/plugin-proposal-object-rest-spread').default,
        {
          useBuiltIns: true
        }
      ],
      [
        require('@babel/plugin-transform-runtime').default,
        {
          helpers: false,
          regenerator: true
        }
      ],
      [
        require('@babel/plugin-transform-regenerator').default,
        {
          async: false
        }
      ]
    ].filter(Boolean)
  }
}

```
</details>

<details>
<summary>tsconfig.json</summary>

```json
{
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [
      "dom",
      "es6",
      "es2017.object",
      "esnext.asynciterable"
    ],
    "module": "esnext",
    "moduleResolution": "node",
    "baseUrl": ".",
    "paths": {
        "@/*": ["app/javascript/*"],
        "#/*": ["app/javascript/modules/*"]
    },
    "sourceMap": true,
    "strict": true,
    "target": "esnext" // target esnext and let babel transpile
  },
  "exclude": [
    "**/*.spec.ts",
    "node_modules",
    "vendor",
    "public"
  ],
  "compileOnSave": false
}
```
</details>

<details>
<summary>app/javascript/spec/tsconfig.json</summary>

```json
{
  "extends": "../../../tsconfig.json",
  "include": [
    "**/*.spec.ts",
    "../../../**/*.d.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}
```
</details>

### Code editor or IDE name and version

Visual Studio Code v1.33.0

### OS name and version

OSX 10.12.6

**Comment by smcenlly** [comment]
It looks like your `given2` import is incorrect. From `given2` [docs](https://www.npmjs.com/package/given2):

**Usage**
---
To use given you just need to require or import the given2 module in your spec files
```javascript
import given from 'given2'
 
describe('Example', () => {
  given('foo', () => 'bar');
 
  it('foo should be "bar"', () => {
    expect(given.foo).toBe('bar');
  })
})
```

I was able to get `given2` using the following steps:

1) Using vue-cli, `create` a new project with the following options:

```
? Please pick a preset: **Manually select features**
? Check the features needed for your project: **Babel, TS, Unit**
? Use class-style component syntax? **Yes**
? Use Babel alongside TypeScript for auto-detected polyfills? **Yes**
? Pick a unit testing solution: **Jest**
? Where do you prefer placing config for Babel, PostCSS, ESLint, etc.? **In dedicated config files**
? Package Manager: **npm**
```

2) Installed `expect2` using:

```
npm install given2 --save-dev
```

3) Update jest configuration to run `given2/setup` file as per `given2` docs:

```
"setupTestFrameworkScriptFile": "given2/setup"
```

4) Updated `example.spec.ts` that was created by the cli with: 

```javascript
import given from 'given2'

describe('Example', () => {
  given('foo', () => 'bar');
 
  it('foo should be "bar"', () => {
    expect(given.foo).toBe('bar');
  })
})
```

At this point, wallaby works fine, but my jest-cli (`npm run test:unit`) is failing with a type error. This is because wallaby ignores type errors, but `test:unit` first compiles your TypeScript code before it starts your tests and fails if TypeScript compilation fails.

```
 tests/unit/example.spec.ts:1:19 - error TS7016: Could not find a declaration file for module 'given2'.
 './node_modules/given2/src/given.js' implicitly has an 'any' type.
```

I couldn't find any types for the given2 library. If you don't have these or can't find them and want `jest` to work outside of wallaby, you'll need to create your own declaration in a `.d.ts` file with the content:

```
declare module 'given2'
```

**Comment by bbugh** [comment]
Hi, thank you for the reply, but no, given2 declaration  is not my issue. I have a custom declaration for given2 (or I couldn’t use it anywhere). My imports work fine with Jest. Jest runs my tests fine. The way I’m importing given2 is the typescript official way to require non default export [see here](https://www.typescriptlang.org/docs/handbook/modules.html). 

I don’t use vue cli3, and their setup is quite different from mine so unfortunately the example isn’t very helpful. Something is wrong with the wallaby config, because everything else works fine. .

**Comment by smcenlly** [comment]
We're unable to reproduce the issue and will require a bit more information to proceed with further investigation. Can you please provide us with:

1) Your jest configuration 
2) Your package.json `dependencies` and `devDependencies`

**Comment by bbugh** [comment]
Thanks for reopening @smcenlly! I apologize for not including the jest config originally, I had it at one point while making the issue and I guess I dropped it while moving things around to fit the template.  Here's the jest config:

```javascript
module.exports = {
  roots: [
    'app/javascript/'
  ],
  moduleDirectories: [
    'node_modules',
    'app/javascript'
  ],
  moduleFileExtensions: [
    'js',
    'json',
    'ts',
    'vue'
  ],
  testRegex: '(/test/.*|\\.(test|spec))\\.(ts|tsx|js)$',
  transform: {
    // '^.+\\.js$': '<rootDir>/node_modules/babel-jest',
    '^.+\\.tsx?$': 'ts-jest',
    '.*\\.(vue)$': 'vue-jest'
  },
  transformIgnorePatterns: [
    '/node_modules/(?!lodash-es)'
  ],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/app/javascript/$1',
    '^#/(.*)$': '<rootDir>/app/javascript/modules/$1'
  },
  snapshotSerializers: [
    '<rootDir>/node_modules/jest-serializer-vue'
  ],
  setupFilesAfterEnv: [
    'given2/setup'
  ],
  setupFiles: [
    'jest-plugin-context/setup'
  ],
  globals: {
    'ts-jest': {
      diagnostics: {
        ignoreCodes: [
          151001 // Ignore the invalid warning about esnext module interop
        ]
      }
    }
  }
}
```

(Note this is Jest 24 which changed `setupTestFrameworkScriptFile` to `setupFilesAfterEnv`)

And the packages from package.json:

```
{
  "dependencies": {
    "@sentry/browser": "^5.0.3",
    "@sentry/integrations": "^5.0.4",
    "actioncable": "^5.2.0",
    "apollo-cache-inmemory": "^1.2.6",
    "apollo-client": "^2.3.8",
    "apollo-link": "^1.2.2",
    "apollo-link-context": "^1.0.8",
    "apollo-link-error": "^1.1.0",
    "apollo-link-http": "^1.5.4",
    "apollo-link-persisted-queries": "^0.2.0",
    "apollo-link-ws": "^1.0.8",
    "apollo-upload-client": "^8.1.0",
    "apollo-utilities": "^1.0.18",
    "bootstrap-vue": "^2.0.0-rc.11",
    "graphql": "^0.13.2",
    "graphql-ruby-client": "^1.4.0",
    "graphql-tag": "^2.9.2",
    "lodash-es": "^4.17.10",
    "nprogress": "^0.2.0",
    "subscriptions-transport-ws": "^0.9.14",
    "vue": "~2.5.17",
    "vue-analytics": "^5.16.0",
    "vue-apollo": "^3.0.0-beta.28",
    "vue-content-loader": "^0.2.1",
    "vue-error-page": "^1.1.6",
    "vue-froala-wysiwyg": "^2.8.5",
    "vue-image-upload-resize": "^1.1.6",
    "vue-intercom": "^1.0.12",
    "vue-mq": "^0.2.1",
    "vue-notification": "^1.3.16",
    "vue-router": "^3.0.1",
    "vue-simple-spinner": "^1.2.8",
    "vue-typed-mixins": "^0.1.0",
    "vuelidate": "^0.7.4",
    "whatwg-fetch": "^2.0.4"
  },
  "devDependencies": {
    "@babel/preset-typescript": "^7.3.3",
    "@rails/webpacker": "4.0.2",
    "@storybook/addon-actions": "^5.0.0",
    "@storybook/addon-knobs": "^3.4.10",
    "@storybook/addon-links": "^5.0.0",
    "@storybook/addons": "^5.0.0",
    "@storybook/vue": "^5.0.0",
    "@types/jest": "^24.0.11",
    "@types/lodash-es": "^4.17.1",
    "@typescript-eslint/parser": "^1.6.0",
    "@vue/test-utils": "^1.0.0-beta.28",
    "babel-core": "^7.0.0-bridge.0",
    "babel-eslint": "^10.0.1",
    "babel-jest": "^24.7.1",
    "babel-loader": "^8.0.5",
    "cache-loader": "^2.0.1",
    "csscomb": "^4.2.0",
    "docdash": "^0.4.0",
    "eslint": "^5.16.0",
    "eslint-loader": "^2.1.2",
    "eslint-plugin-vue": "^5.2.2",
    "fork-ts-checker-webpack-plugin": "^1.0.1",
    "given2": "^2.1.5",
    "husky": ">=1",
    "jest": "^24.7.1",
    "jest-junit": "^6.3.0",
    "jest-plugin-context": "^2.9.0",
    "jest-serializer-vue": "^2.0.2",
    "jsdoc": "^3.5.5",
    "lint-staged": ">=8",
    "node-sass": "^4.11.0",
    "npm": "^6.3.0",
    "sass-lint": "^1.12.1",
    "simple-progress-webpack-plugin": "^1.1.2",
    "thread-loader": "^2.1.2",
    "ts-jest": "^24.0.2",
    "ts-loader": "5.3.3",
    "tslint": "^5.15.0",
    "tslint-eslint-rules": "^5.4.0",
    "typescript": "^3.4.1",
    "vue-jest": "^3.0.4",
    "vue-loader": "^15.3.0",
    "vue-template-compiler": "~2.5.17",
    "webpack-bundle-analyzer": "^2.13.1",
    "webpack-dev-server": "3.2.1",
    "webpack-notifier": "^1.6.0"
  }
}
```

**Comment by smcenlly** [comment]
Thanks for sending through the configuration. We were able to duplicate the problem. Adding a `compilers` section and specifying the module type fixed the problem. Please find the working **wallaby.js configuration** below:

```diff
process.env.BABEL_ENV = 'test'

module.exports = wallaby => {
  return {
    files: [
      'app/javascript/**/*',
      '!app/javascript/**/*.spec.js',
      '!app/javascript/**/*.spec.ts',
      'package.json',
      'jest.config.js',
      'tsconfig.json'
    ],

    tests: ['app/javascript/**/*.spec.js', 'app/javascript/**/*.spec.ts'],

    env: {
      type: 'node',
      runner: 'node'
    },

+   compilers: {
+     '**/*.ts?(x)': wallaby.compilers.typeScript({ 
+       module: "CommonJS"
+      })
+   },

    preprocessors: {
      '**/*.js?(x)': file =>
        require('babel-core').transform(file.content, {
          sourceMap: true,
          compact: false,
          filename: file.path,
          babelrc: true
        })
    },

    setup (wallaby) {
      const jestConfig = require('./package').jest || require('./jest.config')
      delete jestConfig.transform['^.+\\.tsx?$']
      wallaby.testFramework.configure(jestConfig)
    },

    testFramework: 'jest',

    debug: true
  }
}
```

If for some reason this doesn't work for you, can you please provide us with a sample repo. Given that we have duplicated and then fixed the problem with this config, we're not sure what else could be wrong.

**Comment by bbugh** [comment]
Thanks for the idea, but none of my imports are set up for commonjs, and I don't think the module type is going to matter for an import when `allowSyntheticDefaultImports` is disabled and it's a UMD module. 

If I change my `tsconfig.json` to set `allowSyntheticDefaultImports` and `esModuleInterop` both to true, then Jest allows me to `import given from 'given2'` as you suggested should work, but Wallaby still doesn't work (same error).

It seems to me that `allowSyntheticDefaultImports` and `esModuleInterop` are forced `true` on Wallaby, and trying to disable them doesn't have any effect:

```
    compilers: {
      '**/*.ts?(x)': wallaby.compilers.typeScript({
        allowSyntheticDefaultImports: false, 
        esModuleInterop: false
      })
    },
```

(I also tried `true` which didn't work either). 

Does that seem plausible?

**Comment by bbugh** [comment]
I've combed over all of my configurations and everything - babel, jest, webpack, typescript, etc. - seem correct, and they're all behaving correctly. Wallaby is the only thing not behaving as expected with correct code.

I'm unfortunately not able to provide an example repo because this is a huge production app with a complicated setup. I would appreciate continued support since I just bought a license a couple of months ago and I'd like it to not go to waste.

**Comment by bbugh** [comment]
@smcenlly you were right! The `commonjs` module setting worked. It still wasn't working and Wallaby kept dying on the import. I got inspired by my days with Windows and tried a complete restart of everything I could.

I set the `module` in the compiler, then rebooted the computer, delete every webpack, babel, etc. cache I could, uninstalled and reinstalled wallaby extension. For extra emphasis, I smacked the computer too.  Somewhere along in that it started working again.  Thank you so much for your time and effort helping!
